src/nixenv.rs

```diff
@@ -0,0 +1,545 @@
+                                .await
+                                .map_err(|e| {
+                                    NixEnvError::StartNixCommand(
+                                        "nix-env --uninstall'ing a conflicting package",
+                                        e,
+                                    )
+                                })?;
+
+                        if !output.status.success() {
+                            return Err(NixEnvError::NixCommand(
+                                "nix-env --uninstall'ing a conflicting package",
+                                output,
+                            ));
+                        }
+                    }
+                }
+            }
+
+            let output = tokio::process::Command::new(self.nix_store_path.join("bin/nix-env"))
+                .process_group(0)
+                .arg("--profile")
+                .arg(&temporary_profile)
+                .arg("--install")
+                .arg(pkg)
+                .output()
+                .await
+                .map_err(|e| {
+                    NixEnvError::StartNixCommand("Adding a new package to the profile", e)
+                })?;
+
+            if !output.status.success() {
+                return Err(NixEnvError::AddPackage(pkg.to_path_buf(), output));
+            }
+        }
+
+        // Finish by setting the user provided profile to the new version we've constructed
+        {
+            let output = tokio::process::Command::new(self.nix_store_path.join("bin/nix-env"))
+                .process_group(0)
+                .arg("--profile")
+                .arg(self.profile)
+                .arg("--set")
+                .arg(&temporary_profile)
+                .output()
+                .await
+                .map_err(|e| {
+                    NixEnvError::StartNixCommand(
+                        "nix-env --profile ... --set ... the user's profile",
+                        e,
+                    )
+                })?;
+
+            if !output.status.success() {
+                return Err(NixEnvError::UpdateProfile(
+                    self.profile.to_path_buf(),
+                    output,
+                ));
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Debug, serde::Deserialize)]
+struct PackageInfo {
+    #[serde(default)]
+    outputs: HashMap<String, PathBuf>,
+}
+
+fn collect_children<P: AsRef<std::path::Path>>(
+    base_path: P,
+) -> Result<HashSet<PathBuf>, std::io::Error> {
+    let paths = walkdir::WalkDir::new(&base_path)
+        .follow_links(true)
+        .into_iter()
+        .filter_map(|entry| {
+            let entry = entry.ok()?;
+            if entry.file_type().is_dir() {
+                None
+            } else {
+                Some(entry)
+            }
+        })
+        .map(|entry| {
+            entry
+                .path()
+                .strip_prefix(&base_path)
+                .unwrap_or_else(|_| entry.path())
```

Comment:
What's the point of stripping the prefix if we ignore when it fails? Sounds like we'd then have paths like `/nix/store/aaaaaaaaaaaaaa-foo-bar/bin/foo` and `bin/bar`, given a `base_path` of `/nix/store/bbbbbbbbbbbbbbbbb-bar-foo`